rllib/env/tcp_client_inference_env_runner.py

```diff
@@ -0,0 +1,608 @@
+        rewards.append(reward)
+        episode_return += reward
+
+        # We have to create a new episode record.
+        if timesteps == env_steps_per_sample or terminated or truncated:
+            episodes.append(
+                {
+                    Columns.OBS: observations,
+                    Columns.ACTIONS: actions,
+                    Columns.ACTION_DIST_INPUTS: action_dist_inputs,
+                    Columns.ACTION_LOGP: action_logps,
+                    Columns.REWARDS: rewards,
+                    Columns.TERMINATEDS: terminated,
+                    Columns.TRUNCATEDS: truncated,
+                }
+            )
+            # We collected enough samples -> Send them to server.
+            if timesteps == env_steps_per_sample:
+                # Make sure the amount of data we collected is correct.
+                assert sum(len(e["actions"]) for e in episodes) == env_steps_per_sample
+
+                # Send the data to the server.
+                if force_on_policy:
+                    _send_message(
+                        sock_,
+                        {
+                            "type": MessageTypes.EPISODES_AND_GET_STATE.name,
+                            "episodes": episodes,
+                            "timesteps": timesteps,
+                        },
+                    )
+                    # We are forced to sample on-policy. Have to wait for a response
+                    # with the state (weights) in it.
+                    msg_type, msg_body = _get_message(sock_)
+                    assert msg_type == MessageTypes.SET_STATE
+                    onnx_session, output_names = _set_state(msg_body)
+
+                # Sampling doesn't have to be on-policy -> continue collecting
+                # samples.
+                else:
+                    raise NotImplementedError
+
+                timesteps = 0
+
+            # Set new buckets to empty lists (for next episode).
+            observations = [observations[-1]]
+            actions = []
+            action_dist_inputs = []
+            action_logps = []
+            rewards = []
+
+            # The episode is done -> Reset.
+            if terminated or truncated:
+                print(f"Episode done: Return={episode_return}")
+                obs, _ = env.reset()
+                observations = [obs.tolist()]
+                episode_return = 0.0
+
+
+if __name__ == "__main__":
+    from ray.rllib.algorithms.ppo import PPOConfig
+
+    # Create an EnvRunner.
+    config = PPOConfig().environment(
+        observation_space=gym.spaces.Box(-1.0, 1.0, (4,), np.float32),
+        action_space=gym.spaces.Discrete(2),
+        env_config={"port": 5556},
+    )
+    env_runner = TcpClientInferenceEnvRunner(config=config)
+
+    # Create a dummy client playing cartpole.
+    client_thread = threading.Thread(target=_dummy_client)
+    client_thread.start()
+
+    # Query the EnvRunner.
+    episodes = env_runner.sample()
```

Comment:
Nice! But question: In case I want to run with tune I could simply pass in: `TcpClientInferenceEnvRunner` into the `env_runner_class` attribute?

Reply:
Removed this code. It was an early testing leftover.